File: app/src/main/java/com/tomodachi/app/MessageSyncService.java
package com.tomodachi.app;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.webkit.WebViewAssetLoader;

/**
 * ================================================================
 *  خدمة مراقبة الرسائل بالخلفية (حل مجاني بدون Firebase Cloud Messaging)
 * ================================================================
 * تبدأ فقط لما MainActivity يغيب عن الواجهة (راجع MainActivity.onStop) وتبقى
 * تعمل حتى لو المستخدم أغلق التطبيق كلياً من قائمة التطبيقات الأخيرة - لأنها
 * Foreground Service مستقلة عن دورة حياة الـ Activity، ولا نوقفها من
 * onTaskRemoved عمداً.
 *
 * الفكرة: بدل تكرار منطق فايرستور بجافا (يحتاج مكتبة Firestore Android SDK
 * منفصلة + google-services.json جديد)، نحمّل صفحة ويب خفيفة جداً
 * (assets/notify-listener.html) بنفس نطاق appassets.androidplatform.net
 * المستخدم بـ MainActivity، فتشترك بنفس تخزين WebView (localStorage/
 * IndexedDB) وبالتالي بنفس جلسة Firebase Auth المحفوظة - فتقدر تراقب
 * مجموعة "messages" بفايرستور مباشرة بنفس كود جافاسكربت الموجود أصلاً،
 * وتنادي AndroidBridge.showMessageNotification() بالضبط متل ما تفعل
 * MainActivity عادي.
 *
 * حدود هذا الحل (المستخدم اختاره عن قصد بدل FCM المدفوع): نظام أندرويد قد
 * يوقف الخدمة بعد فترة طويلة بالخلفية توفيراً للبطارية (خصوصاً على بعض
 * الهواتف اللي تتشدد بإدارة التطبيقات بالخلفية) - فهو "أفضل جهد" وليس
 * ضمان 100% مثل FCM الحقيقي.
 */
public class MessageSyncService extends Service {

    private static final String ASSET_DOMAIN = "appassets.androidplatform.net";
    private static final String SERVICE_CHANNEL_ID = "tomodachi_bg_sync";
    private static final int SERVICE_NOTIF_ID = 9001;

    private WebView webView;

    @Override
    public void onCreate() {
        super.onCreate();
        NotifyHelper.ensureMessagesChannel(this);
        startForeground(SERVICE_NOTIF_ID, buildServiceNotification());
        setupWebView();
        webView.loadUrl("https://" + ASSET_DOMAIN + "/assets/notify-listener.html");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // START_STICKY: لو النظام قتل العملية لضغط ذاكرة، يحاول إعادة تشغيل
        // الخدمة لاحقاً (بدون نية الـ Intent الأصلية) بدل تركها متوقفة تماماً.
        return START_STICKY;
    }

    /** لا نوقف الخدمة عند إزالة التطبيق من قائمة "الأخيرة" - هذا هو صلب
     *  ميزة "إشعارات حتى لو خرجت من التطبيق" المطلوبة. */
    @Override
    public void onTaskRemoved(Intent rootIntent) {
        super.onTaskRemoved(rootIntent);
        // عمداً: لا شيء هنا (لا نستدعي stopSelf()).
    }

    private void setupWebView() {
        webView = new WebView(getApplicationContext());
        var settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setAllowFileAccess(false);
        settings.setAllowContentAccess(false);
        settings.setMixedContentMode(WebSettings.MIXED_CONTENT_NEVER_ALLOW);

        var assetLoader = new WebViewAssetLoader.Builder()
                .setDomain(ASSET_DOMAIN)
                .addPathHandler("/assets/", new WebViewAssetLoader.AssetsPathHandler(this))
                .build();

        webView.setWebViewClient(new WebViewClient() {
            @Nullable
            @Override
            public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
                return assetLoader.shouldInterceptRequest(request.getUrl());
            }
        });

        webView.addJavascriptInterface(new SyncBridge(), "AndroidBridge");
    }

    private class SyncBridge {
        @android.webkit.JavascriptInterface
        public void showMessageNotification(String title, String body) {
            NotifyHelper.postMessageNotification(MessageSyncService.this, title, body);
        }

        /** تُنادى من notify-listener.html لو ما فيه جلسة مستخدم فعالة، حتى
         *  لا تبقى الخدمة (وإشعارها الدائم بالشريط) شغّالة بلا فائدة. */
        @android.webkit.JavascriptInterface
        public void onListenerIdle(String reason) {
            stopSelf();
        }
    }

    private android.app.Notification buildServiceNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    SERVICE_CHANNEL_ID,
                    "مزامنة الخلفية",
                    NotificationManager.IMPORTANCE_MIN);
            channel.setDescription("يبقي Tomodachi يرصد الرسائل الجديدة أثناء إغلاق التطبيق");
            channel.setShowBadge(false);
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) manager.createNotificationChannel(channel);
        }

        return new NotificationCompat.Builder(this, SERVICE_CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle("Tomodachi")
                .setContentText("جاري رصد الرسائل الجديدة بالخلفية")
                .setPriority(NotificationCompat.PRIORITY_MIN)
                .setOngoing(true)
                .setSilent(true)
                .build();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onDestroy() {
        if (webView != null) {
            webView.destroy();
            webView = null;
        }
        super.onDestroy();
    }
}

File: app/src/main/java/com/tomodachi/app/MessageSyncService.kt
package com.tomodachi.app

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import com.tomodachi.app.data.SessionManager

/**
 * ================================================================
 *  خدمة مراقبة الرسائل بالخلفية - نسخة أصلية بالكامل
 * ================================================================
 * كانت هذه الخدمة سابقاً تحمّل WebView خفي بصفحة notify-listener.html
 * لأنها لم تكن تريد تكرار منطق فايرستور بجافا وقتها. الآن، بما أن التطبيق
 * كله أصبح كوتلن أصلي ومكتبة Firestore Android SDK موجودة أصلاً بالمشروع،
 * لا حاجة لأي WebView إطلاقاً - نراقب مجموعة "messages" مباشرة بنفس مشروع
 * Firebase (semo-chat-f5fdf) ونُصدر إشعاراً حقيقياً عند وصول رسالة جديدة
 * ليست من المستخدم الحالي نفسه.
 *
 * نفس حدود الحل الأصلي: "أفضل جهد" بدون Firebase Cloud Messaging - قد
 * يوقف النظام الخدمة بعد فترة طويلة بالخلفية على بعض الهواتف المتشددة
 * بإدارة البطارية.
 */
class MessageSyncService : Service() {

    private var registration: ListenerRegistration? = null
    private var isFirstSnapshot = true

    override fun onCreate() {
        super.onCreate()
        NotifyHelper.ensureMessagesChannel(this)
        startForeground(SERVICE_NOTIF_ID, buildServiceNotification())
        startListening()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int = START_STICKY

    override fun onTaskRemoved(rootIntent: Intent?) {
        super.onTaskRemoved(rootIntent)
        // عمداً: لا نوقف الخدمة عند إزالة التطبيق من قائمة "الأخيرة".
    }

    private fun startListening() {
        val session = SessionManager(applicationContext)
        val myUsername = session.lastUser
        if (myUsername == null) {
            stopSelf()
            return
        }

        registration = FirebaseFirestore.getInstance()
            .collection("messages")
            .whereEqualTo("deleted", false)
            .orderBy("timestamp", Query.Direction.DESCENDING)
            .limit(1)
            .addSnapshotListener { snapshot, _ ->
                if (isFirstSnapshot) {
                    // أول قراءة عند بدء الخدمة هي الرسالة الحالية الموجودة
                    // مسبقاً - لا نُنبّه عنها، فقط عن أي رسالة تصل بعدها.
                    isFirstSnapshot = false
                    return@addSnapshotListener
                }
                val doc = snapshot?.documents?.firstOrNull() ?: return@addSnapshotListener
                val sender = doc.getString("username") ?: return@addSnapshotListener
                if (sender == myUsername) return@addSnapshotListener

                val isSticker = doc.getBoolean("sticker") == true
                val text = if (isSticker) "🎨 أرسل ستيكر" else (doc.getString("text") ?: "")
                NotifyHelper.postMessageNotification(this, sender, text)
            }
    }

    private fun buildServiceNotification(): Notification {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                SERVICE_CHANNEL_ID,
                "مزامنة الخلفية",
                NotificationManager.IMPORTANCE_MIN
            )
            channel.description = "يبقي Semo يرصد الرسائل الجديدة أثناء إغلاق التطبيق"
            channel.setShowBadge(false)
            val manager = getSystemService(NotificationManager::class.java)
            manager?.createNotificationChannel(channel)
        }

        return NotificationCompat.Builder(this, SERVICE_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentTitle("Semo")
            .setContentText("جاري رصد الرسائل الجديدة بالخلفية")
            .setPriority(NotificationCompat.PRIORITY_MIN)
            .setOngoing(true)
            .setSilent(true)
            .build()
    }

    override fun onBind(intent: Intent): IBinder? = null

    override fun onDestroy() {
        registration?.remove()
        registration = null
        super.onDestroy()
    }

    companion object {
        private const val SERVICE_CHANNEL_ID = "tomodachi_bg_sync"
        private const val SERVICE_NOTIF_ID = 9001
    }
}
